### ISSUE TITLE
Partial application

### ISSUE THREAD

**Issue by wcjohnson** [wcjohnson/lightscript#16]
## Partial application

With the implementation of syntactic placeholders (https://github.com/wcjohnson/lightscript/issues/15) it should now be possible to implement a native version of @citycide's `babel-plugin-partial-application` (see https://github.com/citycide/babel-plugin-partial-application)

The primary advantage here over using the plugin will be that the parsing, linting, and compilation is all under one umbrella, so e.g.:

- Placeholder followed by integer `_1` will be interpreted as a `PlaceholderExpression` with `{index: 1}` rather than as an `Identifier`, enabling a shorter swizzle syntax without introducing false-positive linter errors.

- Placeholder can be configured by the author of a file using Ecmascript directive syntax.

- AST transformation for partial application can be aware of other Lightscript AST transformations, so things like this don't happen:

https://wcjohnson.github.io/lscdiag/?v=1&s=eMOafcKOw4Fuw4IwDMKGXyXDssK5wpMKw5XChsKUE0_DgAsAQl7CmjLCg8KbRMKJwoNAXcOffcOOw67Cmy_CtsKlw4_Cn8O_BSZ6SsONwr7CgF3DoDJjan3Cv8Kfwr3DoMKIwoJtc8K3w6HCsX3DpcO7w5RrDcO3wo_Dtx3DrzANDwErwrnDusK1w7vCj1jCoMOEwprCnW8TFwcWZhTDt2Uuw7YUwozDuTYba8K2wqDCikrCjRDDvMOUIMOHwr7Dq8OPHRTDj8OewokfD3HDlMOzUMKZw5vCqxgmwrrDvsOGwopzIsO2WcKVwozDosKLQAfCicOrwpVCU0DDgizChMO8woYpMTkUwooBw47Dq8KqCiZlwpXDuDvCtXI_wqHDj1zDgA#

**Comment by wcjohnson** [comment]
## Lift Operator

### Motivating Problem
https://wcjohnson.github.io/lscdiag/?v=1&s=eMOafcKQQW_CgzAMwoXDv0rDpBNIwpTCsk50EsKHwqnDmk7Cu8OswrRbwqnCkMKBQMKzwpoQEUfDm8KEw7jDrzPCqMOXLcKHw4R5w7rDtMO8w6wZOsOzw41hw5IewooZwqoBw53DusKeTsKDZmzCkXHDvTXCn8O5MXfDpinDj8OkPMKaY37DiMKvw67Ch245FDwFwr0kw78RM8O4MUzCjV4rw7INFMKww5_Cq8KPwqvDsWrDgMKbw7bCisKlTMOTwrTCtBhVwonCqsKjKsKOS1tawoFeAisfwpwbwr1Wb8OqCy1vcFHDmgjCp8O-IVFyH2LCtXtWeBfDqmjCk8OEACRUMGtPw4ZaRjtnSXZJw4Brw5INw6vDtn1sJcKQDURrw7jDkXbCpsOfBh0HZ0hPEsKSwpDCtWdIwqDDk3hfw48ZasK0w70rEsKBWDkKwr3CscKbw6xwYsKDwrRDw6fDiDTDiGbCtHBZFsKxJiMeQsO8wr0fw6F-AQZnfGY

plugin-partial-application's documented behavior generates a nested function here:
```js
a(_, b(_))
// compiles to
_arg => {
  return a(_arg, _arg2 => {
    return b(_arg2);
  });
};
```
This isn't actually an issue with plugin-partial-application, it's the documented behavior. But what if I don't want the second function to be nested, but rather as the second arg of an outer function? Eg:
```js
(_arg1, _arg2) => a(_arg1, b(_arg2))
```

This is an example of a more general problem, which is that if we implicitly lift expressions to functions whenever we see `_`, and the expression has any depth to it, we have to make arbitrary decisions about the way the expression gets lifted at each node of the expression parse tree -- leaving the user with (1) the need to figure out/remember all the implicit rules and (2) the rigidity of not being able to do something different.

### Solution

Make the implicit explicit; put the user in control of how an expression gets lifted. Introduce a new operator, the "lift operator," which is a unary prefix operator that lifts its RHS into a function, replacing all syntactic placeholders as args. Lift operators can be nested to generate nested functions.

Eg:
```js
<| _
// compiles to
(_arg) -> _arg

<| _.x
// compiles to
(_arg) -> _arg.x

<| f(_, a, _)
// compiles to
(_arg, _arg2) -> f(_arg, a, _arg2)
```

- The expression on the RHS can be any expression here.
- I'm initially proposing the symbol `<|`to represent the lift operator, since it's sort of the opposite of the pipe operator `|>` (eg. `a |> <| _ === a`)

### How does this solve the problem?
```js
<| a(_, <| b(_))
// compiles to
_arg => {
  return a(_arg, _arg2 => {
    return b(_arg2);
  });
};
```
whereas
```js
<| a(_, b(_))
// compiles to
(_arg, _arg2) => a(_arg, b(_arg2))
```

**Comment by wcjohnson** [comment]
The outer `<|` could be considered implicit, with the explicit use of `<|` only required when nesting is desired.

Eg:
```js
a(_, b(_))
// compiles to
(_arg1, _arg2) => a(_arg1, b(_arg2))
```
```js
a(_, <| b(_))
// compiles to
(_arg1) => a(_arg1, (_arg2) => b(_arg2))
```

**Comment by wcjohnson** [comment]
I believe the current implicit-arg design represents the right way to do this stuff. Implemented in 2.3.0. Closing.
